### PR TITLE
memdb: retain old version nodes of ART to satisfy snapshot read

### PR DESCRIPTION
**Pull request by you06**
ref pingcap/tidb#57425

## Changes

The snapshot iterators always read from a snapshot of MemBuffer, but writes between `Next` calls can alter the structure of ART (see the "but explanation" section for details), potentially causing the snapshot iterator to read incorrect results.

This PR introduces a counter for active snapshots. When the counter is greater than 0, it indicates that old versions need to be retained for snapshot reads. In such cases, we store freed nodes in unused slices and delay the actual free operation to prevent them from being reused.

Add a counter for 

## Bug explanation

<details>

<summary>1. The snapshot iterator scans to node (click to expand the figure)</summary>

```text
  │                                 │ 
  │        iterator range           │ 
  ◄─────────────────────────────────► 
  │                                 │ 
  │                                 │ 
  │            ┌────────┐           │ 
  │            │        │           │ 
  │            │  root  │           │ 
  │            │        │           │ 
  │            └────┬───┘           │ 
  │                 │               │ 
  │         ┌───────┴───────┐       │ 
  │         │               │       │ 
  │    ┌────▼───┐      ┌────▼───┐   │ 
  │    │        │      │        │   │ 
  │    │    1   │      │    2   │   │ 
  │    │        │      │        │   │ 
  │    └─────▲──┘      └────────┘   │ 
  │          │                      │ 
  │          │                      │ 
  │          │                      │ 
  │          │                      │ 
  │          │                      │ 
  │    ┌─────┴───┐                  │ 
  │    │ snapshot│                  │ 
  │    │ iterator│                  │ 
  │    └─────────┘                  │ 
  │                                 │ 
```

</details>

<details>

<summary>2. The node 1 grows to larger capacity (node 3) due to coming writes, and the node1 is reused (click to expand the figure)</summary>

```text
 │                                 │                
 │        iterator range           │                
 ◄─────────────────────────────────►                
 │                                 │                
 │                                 │                
 │            ┌────────┐           │                
 │            │        │           │                
 │            │  root  │           │                
 │            │        │           │                
 │            └────┬───┘           │                
 │                 │               │                
 │         ┌───────┴───────┬───────┼─────────┐      
 │         │               │       │         │      
 │    ┌────▼───┐      ┌────▼───┐   │    ┌────▼───┐  
 │    │        │      │        │   │    │        │  
 │    │    3   │      │    2   │   │    │    1   │  
 │    │        │      │        │   │    │        │  
 │    └────────┘      └────────┘   │    └─────▲──┘  
 │                                 │          │     
 │                                 │          │     
 │                                 │          │     
 │                                 │          │     
 │                                 │          │     
 │                                 │    ┌─────┴───┐ 
 │                                 │    │ snapshot│ 
 │                                 │    │ iterator│ 
 │                                 │    └─────────┘ 
 │                                 │                
```

</details>

3. The iterators will return keys out of the given range in following `Next` call.

### CODE REVIEW

File: internal/unionstore/art/art_snapshot.go [hunk 1]
```diff
@@ -91,15 +98,21 @@ func (snap *SnapGetter) Get(ctx context.Context, key []byte) ([]byte, error) {
...
 }
 
 func (i *SnapIter) Value() []byte {
 	return i.value
 }
 
 func (i *SnapIter) Next() error {
+	defer func() {
+		if !i.Valid() {
+			i.counterDec()
```
[the remark this answers is not in the document]
Comment (you06) [reply internal/unionstore/art/art_snapshot.go:113]: Ok, then need little code change in TiDB like `SnapIter.Close` is not called [here](https://github.com/pingcap/tidb/blob/975ad1f403e05652282ce1c07fbd1354f4325b17/pkg/executor/mem_reader.go#L401-L402).

File: internal/unionstore/art/art_snapshot.go [hunk 2]
```diff
@@ -43,12 +43,16 @@ func (t *ART) SnapshotIter(start, end []byte) *SnapIter {
...
 	}
 	it := &SnapIter{
-		Iterator: inner,
-		cp:       t.getSnapshot(),
+		Iterator:   inner,
+		cp:         t.getSnapshot(),
+		isCounting: false,
 	}
 	for !it.setValue() && it.Valid() {
 		_ = it.Next()
```
[the remark this answers is not in the document]
Comment (you06) [reply internal/unionstore/art/art_snapshot.go:51]: The application should read and write concurrently at least, so it's not allowed to write before the iterator is created, unless it'll race.

Will move the `couterInc` before calling `it.Next`.

**Review by ekexium — COMMENTED** [review]
The fix looks OK. I just have some questions:
1. In your example, which variable in the iterator is pointing to node-1? Is it `nodes` in `baseIter`?
2. Why are we only fixing for SnapshotIter? What about the ART Iterator? Is it currently safe in implementation but not guaranteed by design?
3. Are there any other inner structure change that could lead to iterator invalidation, other than the free nodes? I don't see other cases at first glance. Have you verified this as well?

**Comment by you06** [comment]
> * In your example, which variable in the iterator is pointing to node-1? Is it `nodes` in `baseIter`?

Yes, `nodes[len(nodes) - 1]` is `node1` in the example.

> * Why are we only fixing for SnapshotIter? What about the ART Iterator?

The `SnapshotIter` can be explained always read from the snapshot, so we need to protect it against the later writes.

For `Iterator`, there should be no writes during the iterator, unless the result makes no sense.

> Is it currently safe in implementation but not guaranteed by design?

Yes, such usage in TiDB is out of my expectation. For long-term, we may deprecate the `SnapshotIter` and replace it with `SnapshotScan` which returns all the rows in one call.

> * Are there any other inner structure change that could lead to iterator invalidation, other than the free nodes? I don't see other cases at first glance. Have you verified this as well?

I don't see other cases also. [`GetSnapshotValue`](https://github.com/tikv/client-go/blob/3eb6e787c01b1b097bf92bc8c6196339f317f237/internal/unionstore/arena/arena.go#L325-L333) can filter out the new added keys or versions.

**Comment by cfzjywxk** [comment]
@you06 
Please also ensure the `read-with-write` test cases in covered in the PR for tidb repo.

**Review by ekexium — APPROVED** [review]

**Pull request merged**